### PR TITLE
Coding-team cost guardrails: LLM-call budget, scaled round ceiling, per-role thinking

### LINKED ISSUE #772: Coding team: cost guardrails — per-job LLM call budget, sane max_rounds with explicit failure, per-role thinking levels
## Problem

When the coding team misbehaves, the failure mode is unbounded cloud spend followed by a silent success claim:

- The swarm loop (`coding_team/orchestrator.py:345-366`) has **no sleep/backoff and no LLM-call budget**; `max_rounds=500` is the only ceiling. One stuck task costs ~4–15 LLM calls per round (implement session + lint + code review + tech-lead review), back-to-back.
- After 500 fruitless rounds, the orchestrator reports **`status="completed"`, `"Completed: 0 tasks merged"`** (`orchestrator.py:159-164`) — the plain `POST /run` path never downgrades this (only the GitHub hook does, via `_has_merged_tasks`, `api/main.py:1050`).
- Every coding-team agent — including mechanical JSON steps like assignment and the review gates — runs at the model's **maximum thinking level** (the `LLM_ENABLE_THINKING` default) with `max_tokens` defaulting to 32768 (`llm_service/clients/ollama.py:1397-1406`).
- The expensive degenerate case: max-effort thinking consumes the whole budget and returns zero content, which raises `LLMTemporaryError` (`ollama.py:660-670`) and is retried up to `LLM_MAX_RETRIES=10` — each retry again at max thinking. Observed in production logs: *"LLM produced reasoning tokens but no content — model likely spent its token budget on thinking."*

The Strategy Lab already solved this class of problem with `STRATEGY_LAB_DESIGN_MAX_LLM_CALLS` (per-cycle hard budget, `failed: budget_exhausted`); the coding team has no equivalent.

## Proposed fix

- **Per-job LLM-call budget** for the coding team (env-tunable, mirroring the Strategy Lab pattern): charge before every LLM call across Tech Lead, Senior SWEs, and gate agents; on exhaustion terminate with an explicit `failed: budget_exhausted` status.
- **Sane round ceiling**: derive `max_rounds` from task count (e.g. 3× tasks) instead of a flat 500, and make exhaustion an explicit `failed: max_rounds_exhausted` status on every path — never `completed` with zero merges.
- **Per-role thinking levels**: low/no thinking for assignment, lint-fix, and review-gate calls; reserve high thinking for implementation. Make this configurable per model key (the `llm_service` already supports per-call `think` and `LLM_THINKING_LEVEL`).
- **Empty-content retry hygiene**: when a call finishes with reasoning-only/zero content, retry once with reduced thinking (or `think=False`) rather than burning the full transport retry count at max effort.

## Acceptance criteria

- A job whose budget trips ends with an explicit budget-exhausted failure status visible via `GET /status/{job_id}` (and the GitHub issue comment on the run-from-github path); no further LLM calls are made after the trip (regression test with a stub LLM counting calls).
- A swarm that exhausts its round ceiling reports failure — `"completed: 0 tasks merged"` can no longer occur (regression test).
- Assignment/gate calls run at the configured reduced thinking level; implementation calls keep the configured high level (assertable via the LLM client's request parameters in tests).
- New env vars documented in CLAUDE.md's environment-variable table.
- 90% line coverage on new/modified code.

### PR DESCRIPTION
**Pull request by brandonkindred**
Closes #772

## Problem

When the coding-team swarm misbehaves, the failure mode was unbounded cloud spend followed by a silent success claim: the loop ran to a flat `max_rounds=500` with no LLM-call budget, then reported `status="completed"`, `"Completed: 0 tasks merged"`, and every agent — including mechanical JSON steps — ran at the model's max thinking level.

## What's here

Three deterministic guardrails, mirroring the Strategy Lab `LLMCallBudget` pattern. New module `backend/agents/coding_team/_guardrails.py` (stdlib-only, no import cycle).

### 1. Per-job LLM-call budget
- `CodingTeamLLMBudget` is bound for the whole job via `use_budget(...)` and charged before **every** logical model call through a transparent `_BudgetedClient` wrapper around the backing `LLMClient` — one chokepoint covering Tech Lead, Senior SWE (including each tool-loop turn), and the lint/code-review gates. A transport-level retry inside one call charges once.
- `CodingTeamBudgetExhausted` is a **`BaseException`**, not `Exception`, so it propagates straight through the agents' broad `except Exception` handlers (which would otherwise swallow it) to the orchestrator's explicit handler → `status="failed: budget_exhausted"`. **No further LLM calls run after the trip** (pre-charge/check-then-increment).
- Env: `CODING_TEAM_MAX_LLM_CALLS` (default 300, floor 1, garbage→default).

### 2. Task-count-scaled round ceiling with explicit failure
- `max_rounds_for(num_tasks) = max(CODING_TEAM_MIN_ROUNDS, CODING_TEAM_ROUND_MULTIPLIER × tasks)` replaces the flat 500.
- `swarm.run()` now returns an explicit terminal reason (`complete` / `cancelled` / `max_rounds_exhausted`); the orchestrator maps it through `terminal_status()`:
  - exhausted ceiling → `failed: max_rounds_exhausted`
  - complete with zero merges → `failed: no_tasks_merged`
  - complete with merges → `completed`
  - cancelled → left as the swarm set it
- `"completed: 0 tasks merged"` can no longer be produced on **any** path.
- Env: `CODING_TEAM_ROUND_MULTIPLIER` (default 3), `CODING_TEAM_MIN_ROUNDS` (default 10).

### 3. Per-role thinking levels
- Mechanical agents — Tech Lead assignment + the per-task lint and code-review gates — run with `think=False`; implementation and planning keep the configured high level. The Tech Lead builds its assignment/review agents from a separate reduced-thinking model so planning is unaffected.
- Operator override: `CODING_TEAM_THINKING_MECHANICAL` re-enables reasoning everywhere.
- Implemented by extending the shared `llm_service.strands_provider.get_strands_model` to forward a per-instance `think` (cache key now includes it).

### Docs + tests
- Four new env vars documented in `CLAUDE.md`.
- `backend/agents/coding_team/tests/test_cost_guardrails.py` (26 tests, env-independent — no job-service/Postgres fixtures): budget trip with a call-counting stub asserting **zero** calls after the trip; round exhaustion → `failed: max_rounds_exhausted` and proof the false-success string can't appear; `max_rounds` scaling + env overrides/floors/garbage; per-role `think` asserted via the constructed models' config; `get_strands_model` think-forwarding + cache-key; orchestrator terminal-status state machine (completed / max_rounds_exhausted / no_tasks_merged / budget_exhausted / cancelled / default-getter path).
- `_guardrails.py` at **100%** line coverage; orchestrator's changed lines covered (remaining misses are pre-existing swarm internals).

## Acceptance criteria
- [x] Budget trip ends with `failed: budget_exhausted` via `GET /status/{job_id}`; no further LLM calls after the trip (regression test with a call-counting stub).
- [x] A swarm that exhausts its round ceiling reports failure — `"completed: 0 tasks merged"` can no longer occur (regression test).
- [x] Assignment/gate calls run at the reduced thinking level; implementation keeps the high level (asserted via the models' request params in tests).
- [x] New env vars documented in CLAUDE.md's environment-variable table.
- [x] 90% line coverage on new/modified code.

## Follow-up (split out)
The shared empty-content / reasoning-only retry hygiene (retry once with reduced thinking instead of burning `LLM_MAX_RETRIES` at max effort, in `llm_service/clients/ollama.py`) is tracked separately in #774 so the shared retry-path change can be reviewed on its own.

## Notes
- The SE integration path passes `get_llm=llm_service.get_client` (a raw client); the orchestrator now normalises whatever `get_llm` yields to a backing client and rebuilds a budgeted, think-tagged Strands model, so both the default and SE paths are budgeted.
- Two pre-existing test issues in this area are unrelated to this change and left as-is: a stale `test_tech_lead_plan_to_graph.py` that calls `TechLeadAgent(llm=...)`, and `test_strands_provider.py` importing a symbol no longer in `strands_provider.py`.

---
_Generated by [Claude Code](https://claude.ai/code/session_01P2GzwG9z5QLLbb32C3pc2J)_

### CODE REVIEW

File: backend/agents/coding_team/orchestrator.py [hunk 1]
```diff
@@ -108,60 +137,83 @@ def _persist_graph() -> None:
...
     merged_count = sum(1 for t in graph.get_tasks() if t.status == TaskStatus.MERGED)
-    _update(
-        status="completed",
-        phase="completed",
-        status_text=f"Completed: {merged_count} tasks merged",
-    )
+    decision = terminal_status(run_reason, merged_count, run_max_rounds)
+    if decision is not None:
+        status, status_text = decision
+        _update(status=status, phase="completed", status_text=status_text)
```
[the remark this answers is not in the document]
Comment (brandonkindred) [reply backend/agents/coding_team/orchestrator.py]: Good catch — fixed in bcdfc7d. `run_coding_team_orchestrator` now **returns** its terminal status (`completed` / `failed: budget_exhausted` / `failed: max_rounds_exhausted` / `failed: no_tasks_merged`, or `None` for cancelled), and both integration callers branch on it instead of unconditionally marking the job completed:

- `software_engineering_team/orchestrator.py` (the `use_coding_team=True` path) — `failed*` → `JOB_STATUS_FAILED`, `None` → leave the cancelled status, else `JOB_STATUS_COMPLETED`.
- `software_engineering_team/temporal/activities.py:execute_coding_team_activity` — same mapping.

So a guardrail failure (or a cancellation) can no longer be overwritten as a successful SE job. The orchestrator's return value is asserted for every terminal path in `test_cost_guardrails.py`.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by brandonkindred** [comment]
@codex review and tell me the first 20 logic errors that you find

File: backend/agents/coding_team/orchestrator.py [hunk 2]
```diff
@@ -108,60 +149,85 @@ def _persist_graph() -> None:
...
-    )
+        return "failed: budget_exhausted"
 
     merged_count = sum(1 for t in graph.get_tasks() if t.status == TaskStatus.MERGED)
-    _update(
-        status="completed",
-        phase="completed",
-        status_text=f"Completed: {merged_count} tasks merged",
-    )
+    decision = terminal_status(run_reason, merged_count, run_max_rounds)
```
[the remark this answers is not in the document]
Comment (brandonkindred) [reply backend/agents/coding_team/orchestrator.py:225]: Fixed in 3022df1. The run-from-github hook now captures `run_coding_team_orchestrator`'s return and, on a `failed:*` status, records the failure and returns **before** `_has_merged_tasks` / fast-forward / push / PR / `status="completed"`:

```python
ct_status = run_coding_team_orchestrator(...)
...
if ct_status is not None and ct_status.startswith("failed"):
    _record_failure(client, owner, repo, num, job_id, f"coding team {ct_status}")
    return
```

So a budget trip (or `max_rounds_exhausted`) with *partial* merges no longer publishes the partial branch or reports success — the active-issue marker is retained so a retry continues from `development`. Regression test added (`test_retained_when_guardrail_failed_with_partial_merge`): a stub that merges a task **and** returns `failed: budget_exhausted` must not publish. (`None`/legacy returns are untouched, so the existing publish path and its tests are unaffected.)

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by brandonkindred** [comment]
@codex review and tell me the first 20 logic errors that you find

File: backend/agents/coding_team/api/main.py [hunk 3]
```diff
@@ -1047,6 +1047,13 @@ def _run_with_github_hooks(
             _record_failure(client, owner, repo, num, job_id, str(e))
             return
 
+        # A guardrail failure (budget / round ceiling / no merges) must stop
+        # the hook before it publishes — even if some tasks merged, a partial
+        # branch must never be fast-forwarded, pushed, or reported as success.
+        if ct_status is not None and ct_status.startswith("failed"):
```
[the remark this answers is not in the document]
Comment (brandonkindred) [reply backend/agents/coding_team/api/main.py]: Fixed in fc9d455. The hook now treats a `None` return (cancelled) as a non-publishing terminal path, returning before `_has_merged_tasks` / fast-forward / push / PR:

```python
ct_status = run_coding_team_orchestrator(...)
...
if ct_status is None:
    # Cancelled — status already recorded via the callback; leave work
    # unpublished so a retry can resume.
    return
if ct_status.startswith("failed"):
    _record_failure(client, owner, repo, num, job_id, f"coding team {ct_status}")
    return
```

So a cancellation after a partial merge no longer publishes. I also aligned the test orchestrator stubs that simulate completion to return `"completed"` (matching the real contract — `None` only ever means cancelled), and added `test_retained_when_cancelled_with_partial_merge`: a stub that merges a task and returns `None` must not publish (marker retained, `created_pulls == []`). Full `test_github_source.py` + `test_cost_guardrails.py` suite green (100 passed).

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by brandonkindred** [comment]
@codex review and tell me the first 20 logic errors that you find

**Comment by brandonkindred** [comment]
After deeper consideration, I decided that this isn't necessary at this time, nor is it the right solution for managing budget. When budget management is added it should work for the entire system instead of just one team. Closing this PR.
